File: code/utils/find-graph-components.ts
import { Edge, Components, Graph } from "./definitions";
import createDirectedGraph, { createUndirectedGraph } from "./graphs";

export default function FindGraphComponents(edges: Edge[]): Components[] {
    const undirectedGraph = createUndirectedGraph(edges, true);
    const graphs = undirectedGraph.graphs;
    const nodeSet = undirectedGraph.nodes;

    const results = new Array<Components>();
    const visitedSet = new Set<string>();

    nodeSet.forEach(function(node) {
        if (visitedSet.has(node)) {
            return;
        }
        const curComponent: string[] = [ node ];
        visitedSet.add(node);

        const queue = [ node ];
        while (queue.length > 0) {
            // do BFS
            const curQueueSize = queue.length;
            for (let i = 0; i < curQueueSize; i++) {
                const fromNode = queue[0];
                const nextNodes = graphs.get(fromNode);
                queue.shift();
                if (nextNodes !== undefined) {
                    nextNodes.forEach(function(value) {
                        if (visitedSet.has(value)) {
                            return;
                        }
                        visitedSet.add(value);
                        curComponent.push(value);
                        queue.push(value);
                    });
                }
            }
        }
        results.push({ nodes: curComponent });
    });
    
    return results;
}

export function FindComponentsGraph(edges: Edge[]): Graph[] {

    const undirectedGraph = createUndirectedGraph(edges, true);
    const graphs = undirectedGraph.graphs;
    const nodeSet = undirectedGraph.nodes;

    const results = new Array<Graph>();
    const visitedSet = new Set<string>();

    const directedGraph = createDirectedGraph(edges);

    nodeSet.forEach(function(node) {
        if (visitedSet.has(node)) {
            return;
        }
        const curComponent: string[] = [ node ];
        visitedSet.add(node);

        const queue = [ node ];
        while (queue.length > 0) {
            // do BFS
            const curQueueSize = queue.length;
            for (let i = 0; i < curQueueSize; i++) {
                const fromNode = queue[0];
                const nextNodes = graphs.get(fromNode);
                queue.shift();
                if (nextNodes !== undefined) {
                    nextNodes.forEach(function(value) {
                        if (visitedSet.has(value)) {
                            return;
                        }
                        visitedSet.add(value);
                        curComponent.push(value);
                        queue.push(value);
                    });
                }
            }
        }

        const curComponentEdges: Edge[] = [];
        for (let i = 0; i < curComponent.length; i++) {
            for (let j = 0; j < curComponent.length; j++) {
                if (i === j) {
                    continue;
                }

                // console.log(curComponent[i], '->', curComponent[j]);
                const fromNode = curComponent[i];
                const toNode = curComponent[j];
                const fromSet = directedGraph.graphs.get(fromNode);

                if (fromSet === undefined) {
                    continue;
                }

                if (fromSet.has(toNode)) {
                    curComponentEdges.push({ from: fromNode, to: toNode });
                }
            }
        }
        results.push({ edges: curComponentEdges });
    });
    
    return results;
}

// TEST CASE:
// FindGraphComponents([
//     { from: '1', to: '2' },
//     { from: '2', to: '3' },
//     { from: '3', to: '1' },
//     { from: '4', to: '5' },
//     { from: '5', to: '6' },
//     { from: '6', to: '4' },
//     { from: '3', to: '4' },
// ]).forEach(edge => console.log('result: ', edge));

// FindComponentsGraph([
//     { from: '1', to: '2' },
//     { from: '2', to: '3' },
//     { from: '3', to: '1' },
//     { from: '4', to: '5' },
//     { from: '5', to: '6' },
//     { from: '6', to: '4' },
//     // { from: '1', to: '4' },
// ]).forEach(comp => {
//     comp.edges.forEach(edge => console.log(edge))
//     console.log('\nnext\n');
// });

File: code/find_graph_components.py
import graphs
from definitions import Edge, Graph
from collections import deque

def FindComponentsGraph(edges):
    un_graph_result = graphs.createUndirectedGraph(edges, True)
    un_graph = un_graph_result.graphs
    un_nodes = un_graph_result.nodes

    result = []
    visited_set = set()

    di_graph = graphs.createDirectedGraph(edges).graphs

    for node in un_nodes:

        if node in visited_set:
            continue
        visited_set.add(node)

        cur_component = [ node ]
        que = deque([ node ])
        while que: # que in not empty.
            cur_queue_size = len(que)
            for i in range(cur_queue_size):
                from_node = que.popleft()
                next_nodes = un_graph[from_node]
                for next_node in next_nodes:
                    if next_node in visited_set:
                        continue
                    
                    visited_set.add(next_node)
                    cur_component.append(next_node)
                    que.append(next_node)

        cur_components_edges = []

        for i, from_node in enumerate(cur_component):
            for j, to_node in enumerate(cur_component):
                if i == j:
                    continue

                if from_node in di_graph:
                    from_set = di_graph[from_node]
                    if to_node in from_set:
                        cur_components_edges.append(Edge(from_node, to_node))

        result.append(Graph(cur_components_edges))

    return result

# result = FindComponentsGraph([
#     Edge('1', '2'),
#     Edge('2', '3'),
#     Edge('1', '3')
# ])

# for edge in result[0].edges:
#     print('from %s to %s' % (edge.src, edge.dest))
